features/cellular/framework/targets/UBLOX/N2XX/UBLOX_N2XX_CellularStack.cpp

```diff
@@ -0,0 +1,264 @@
+    bool success = true;
+    char ipAddress[NSAPI_IP_SIZE];
+    char *tmpBuf = (char *) malloc(size*2+1);
+    memset(tmpBuf, 0, size*2+1);
+    int port = 0;
+    Timer timer;
+
+    if (socket->pending_bytes == 0) {
+        _at.process_oob();
+        if (socket->pending_bytes == 0) {
+            free(tmpBuf);
+            return NSAPI_ERROR_WOULD_BLOCK;
+        }
+    }
+
+    timer.start();
+    while (success && (length > 0)) {
+        read_blk = N2XX_MAX_PACKET_SIZE;
+        if (read_blk > length) {
+            read_blk = length;
+        }
+        if (socket->pending_bytes > 0) {
+            _at.cmd_start("AT+NSORF=");
+            _at.write_int(socket->id);
+            _at.write_int(read_blk);
+            _at.cmd_stop();
+
+            _at.resp_start();
+            _at.skip_param(); // receiving socket id
+            _at.read_string(ipAddress, sizeof(ipAddress));
+            port = _at.read_int();
+            usorf_sz = _at.read_int();
+
+            // Must use what +NSORF returns here as it may be less or more than we asked for
+            if (usorf_sz >= socket->pending_bytes) {
+                socket->pending_bytes = 0;
+            } else {
+                socket->pending_bytes -= usorf_sz;
+            }
+
+            if (usorf_sz > length) {
+                usorf_sz = length;
+            }
+            _at.read_string(tmpBuf + count, usorf_sz*2+1);
+            _at.resp_stop();
+
+            if (usorf_sz > 0) {
+                count += (usorf_sz*2);
+                length -= (usorf_sz*2);
+            } else {
+                // read() should not fail
+                success = false;
+            }
+        }  else if (timer.read_ms() < SOCKET_TIMEOUT) {
+            // Wait for URCs
+            _at.process_oob();
+        } else {
+            if (count == 0) {
+                // Timeout with nothing received
+                nsapi_error_size = NSAPI_ERROR_WOULD_BLOCK;
+                success = false;
+            }
+            length = 0; // This simply to cause an exit
+        }
+    }
+    timer.stop();
+
+    socket->pending_bytes = 0;
+    if (!count || (_at.get_last_error() != NSAPI_ERROR_OK)) {
+        free(tmpBuf);
+        return NSAPI_ERROR_WOULD_BLOCK;
+    } else {
+        nsapi_error_size = count;
+    }
+
+    if (success && socket->proto == NSAPI_UDP && address) {
+        address->set_ip_address(ipAddress);
+        address->get_ip_address();
+        address->set_port(port);
+    }
+
+    hex_str_to_char_str(tmpBuf, count, (char*) buffer);
+
+    free(tmpBuf);
+    return nsapi_error_size = count/2;
+}
+
+nsapi_error_t UBLOX_N2XX_CellularStack::socket_close_impl(int sock_id)
+{
+    _at.lock();
+    _at.cmd_start("AT+NSOCL=");
+    _at.write_int(sock_id);
+    _at.cmd_stop_read_resp();
+
+    return _at.unlock_return_error();
+}
+
+// Find or create a socket from the list.
+UBLOX_N2XX_CellularStack::CellularSocket *UBLOX_N2XX_CellularStack::find_socket(int id)
```

Comment:
There is a brand new implementation for this `CellularSocket *find_socket(int sock_id)`